manual/manual/tutorials/coreexamples.etex

```diff
@@ -580,6 +580,80 @@ let fixpoint f x =
 the function "f" cannot raise a "Done" exception, which removes an
 entire class of misbehaving functions.
 
+\section{Lazy expressions}
+
+OCaml allows us to defer some computation, usually expensive one, until later when 
+we need the result of that computation. 
+
+We use "lazy (expr)" to delay the evaluation of some expression "expr". For 
+example, we can defer the computation of "1+1" until we need the result of that
+expression, i.e. "2". Let us see how we initialize a lazy expression. 
+
+\begin{caml_example}{toplevel}
+  let lazy_two = lazy ( print_endline "lazy_two evaluation"; 1 + 1 );;
+\end{caml_example}
+
+We added "print_endline \"lazy_two evaluation\"" to see when the lazy
+ expression is being evaluated.
+
+There are several things to note about the result printed by the toplevel above. 
+First, "lazy_two" has type "int lazy_t". However, "'a lazy_t" is a type 
+constructor for keyword "lazy" used by the compiler. 
+One should use type "'a Lazy.t" to work with lazy expressions. Second, the value 
+of "lazy_two" is "<lazy>", which means the expression is not evaluated yet, and 
+the result is unknown.
+
+When we finally need the result of a lazy expression, we can call "Lazy.force"  
+on that expression to force its evaluation. "force" is a function from a 
+standard-library  
+\href{https://caml.inria.fr/pub/docs/manual-ocaml/libref/Lazy.html}{module "Lazy"}.
+
+\begin{caml_example}{toplevel}
+  Lazy.force lazy_two;;
+\end{caml_example}
+
+Notice that our function call above prints ``lazy_two evaluation'' and then 
+returns the plain value of the computation. 
+
+Now if we look at the 
+value of "lazy_two", we see that it is not "<lazy>" anymore but "lazy 2".
+
+\begin{caml_example}{toplevel}
+  lazy_two;;
+\end{caml_example}
+
+This is because "Lazy.force" memoizes the result of the forced expression. In other 
+words, every subsequent call of "Lazy.force" on that expression returns the 
+result of the first computation without recomputing the lazy expression. Let us 
+force "lazy_two" once again. 
+
+\begin{caml_example}{toplevel}
+  Lazy.force lazy_two;;
+\end{caml_example}
+
+The expression is not evaluated this time; notice that "print_endline" is not called. 
+The result of the initial computation is simply returned. 
+
+Lazy patterns provide another way to force a lazy expression. 
+
+\begin{caml_example}{toplevel}
+  let lazy_l = lazy ([1; 2] @ [3; 4]);;
+  let lazy l = lazy_l;;
+\end{caml_example}
+
+We can also use lazy patterns in pattern matching.
+
+\begin{caml_example}{toplevel}
+  let lazy_match lazy_expr = 
+    match lazy_expr with
+    | lazy 0 -> "matches if (Lazy.force lazy_expr) = 0"
+    | lazy _ -> "despite being a lazy wildcard pattern, always forces the lazy expr";;
+\end{caml_example}
```

Comment:
I think it would make the example stronger, if the `lazy _` pattern really forced a computation. Currently, in this example, the computation is already forced by `lazy 0`. A possibility might be
```OCaml
let maybe_eval guard lazy_expr = match guard, lazy_expr with
| lazy false, _ -> "matches if (Lazy.force guard = false)"
| lazy true, lazy _ -> "despite being a lazy wildcard pattern, always forces the lazy expr";;
```
Then the explanation below could mention that the lazy expression is forced only if the `guard` value yields `true` once computed.